File: 백준/Gold/9252. LCS 2/LCS 2.java
import java.io.*;
import java.util.*;

public class Main {
    static int x, y;
    static int[][] data;
    static String result = "";


    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String a = br.readLine();
        String b = br.readLine();

        data = new int[a.length() + 1][b.length() + 1];
        for(int i = 0; i < a.length(); i++){
            Arrays.fill(data[i], 0);
        }

        for(int i = 1; i < a.length() + 1; i++){
            for(int j = 1; j < b.length() + 1; j++){
                if(a.charAt(i - 1) == b.charAt(j - 1)){
                    data[i][j] = data[i - 1][j - 1] + 1;
                }else{
                    data[i][j] = Math.max(data[i - 1][j], data[i][j - 1]);
                }
            }
        }

//        for(int i = 0; i < a.length() + 1; i++) {
//            for (int j = 0; j < b.length() + 1; j++) {
//                System.out.print(data[i][j] + "\t");
//            }
//            System.out.println();
//        }

        x = a.length();
        y = b.length();

        while(x > 0 && y > 0){
            if(data[x][y] == data[x - 1][y]){

                x--;

            }else if(data[x][y] == data[x][y - 1]){

                y--;
            }else{
                result = a.charAt(x - 1) + result;

                x--;
                y--;
            }
        }
        System.out.println(result.length());
        if(result.length() >= 0) {
            System.out.println(result);
        }
    }
}

File: 백준/Gold/9252. LCS 2/LCS 2.py
import sys
input = sys.stdin.readline 

a = input().strip()
b = input().strip()

data = [[0] * (len(b) + 1) for _ in range(len(a) + 1)]

for i in range(1, len(a) + 1):
  for j in range(1, len(b) + 1):
    if(a[i - 1] == b[j - 1]):
      data[i][j] = data[i - 1][j - 1] + 1
    else:
      data[i][j] = max(data[i - 1][j], data[i][j - 1])

result = ""
x, y = len(a), len(b)
while(x > 0 and y > 0):
  if(data[x][y] == data[x - 1][y]):
    x -= 1
  elif(data[x][y] == data[x][y -1]):
    y -= 1
  else:
    result = a[x - 1] + result 
    x -= 1
    y -= 1

length = len(result)
print(length)
if(length != 0):
  print(result)